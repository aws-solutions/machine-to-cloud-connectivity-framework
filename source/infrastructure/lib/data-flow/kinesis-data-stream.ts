// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { KinesisStreamsToKinesisFirehoseToS3 } from '@aws-solutions-constructs/aws-kinesisstreams-kinesisfirehose-s3';
import { Bucket } from 'aws-cdk-lib/aws-s3';
import { NagSuppressions } from 'cdk-nag';
import { Construct } from 'constructs';

export interface KinesisDataStreamConstructProps {
  readonly s3LoggingBucket: Bucket;
}

/**
 * Creates a Kinesis data stream, a Kinesis data firehose, a data S3 bucket, and related roles.
 */
export class KinesisDataStreamConstruct extends Construct {
  public kinesisStreamName: string;
  public dataBucketName: string;

  constructor(scope: Construct, id: string, props: KinesisDataStreamConstructProps) {
    super(scope, id);

    const kinesisStreamsToKinesisFirehoseToS3 = new KinesisStreamsToKinesisFirehoseToS3(this, 'DataStream', {
      existingLoggingBucketObj: props.s3LoggingBucket,
      bucketProps: {
        serverAccessLogsPrefix: 'm2c2data/'
      }
    });
    this.kinesisStreamName = kinesisStreamsToKinesisFirehoseToS3.kinesisStream.streamName;
    this.dataBucketName = (<Bucket>kinesisStreamsToKinesisFirehoseToS3.s3Bucket).bucketName;

    // cdk-nag suppressions
    NagSuppressions.addResourceSuppressions(
      kinesisStreamsToKinesisFirehoseToS3,
      [
        { id: 'AwsSolutions-IAM5', reason: 'It specifies a bucket, so it is not a wildcard permission.' },
        { id: 'AwsSolutions-KDF1', reason: 'The data stream is encrypted.' }
      ],
      true
    );
  }
}
