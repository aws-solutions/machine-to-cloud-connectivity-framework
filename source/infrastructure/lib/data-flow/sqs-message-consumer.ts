// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { IotToSqs, IotToSqsProps } from '@aws-solutions-constructs/aws-iot-sqs';
import { LambdaToDynamoDB } from '@aws-solutions-constructs/aws-lambda-dynamodb';
import { SqsToLambda } from '@aws-solutions-constructs/aws-sqs-lambda';
import { CfnDeletionPolicy, Duration } from 'aws-cdk-lib';
import { AttributeType, CfnTable, Table } from 'aws-cdk-lib/aws-dynamodb';
import { Code, Runtime } from 'aws-cdk-lib/aws-lambda';
import { IBucket } from 'aws-cdk-lib/aws-s3';
import { NagSuppressions } from 'cdk-nag';
import { Construct } from 'constructs';

export interface SQSMessageConsumerConstructProps {
  readonly solutionConfig: {
    loggingLevel: string;
    solutionId: string;
    solutionVersion: string;
    sourceCodeBucket: IBucket;
    sourceCodePrefix: string;
  };
}

/**
 * Creates a message SQS queue, a message consumer Lambda function, a logs DynamoDB tables, and IoT rules.
 */
export class SQSMessageConsumerConstruct extends Construct {
  public logsTable: Table;
  private readonly sql: string = `SELECT topic(3) as connectionName, topic(2) as logType, timestamp() as timestamp, * FROM 'm2c2/{type}/+'`;

  constructor(scope: Construct, id: string, props: SQSMessageConsumerConstructProps) {
    super(scope, id);

    const sourceCodeBucket = props.solutionConfig.sourceCodeBucket;
    const sourceCodePrefix = props.solutionConfig.sourceCodePrefix;

    const sqsToLambda = new SqsToLambda(this, 'SQSMessageConsumer', {
      lambdaFunctionProps: {
        code: Code.fromBucket(sourceCodeBucket, `${sourceCodePrefix}/sqs-message-consumer.zip`),
        description: 'Machine to Cloud Connectivity SQS message consumer function',
        environment: {
          LOGGING_LEVEL: props.solutionConfig.loggingLevel,
          SOLUTION_ID: props.solutionConfig.solutionId,
          SOLUTION_VERSION: props.solutionConfig.solutionVersion
        },
        handler: 'sqs-message-consumer/index.handler',
        retryAttempts: 0,
        runtime: Runtime.NODEJS_14_X,
        timeout: Duration.minutes(1)
      },
      queueProps: {
        visibilityTimeout: Duration.minutes(1)
      },
      maxReceiveCount: 3
    });

    const lambdaToDynamoDb = new LambdaToDynamoDB(this, 'LogsDynamoDB', {
      existingLambdaObj: sqsToLambda.lambdaFunction,
      dynamoTableProps: {
        partitionKey: {
          name: 'connectionName',
          type: AttributeType.STRING
        },
        sortKey: {
          name: 'timestamp',
          type: AttributeType.NUMBER
        },
        timeToLiveAttribute: 'ttl'
      },
      tableEnvironmentVariableName: 'LOGS_DYNAMODB_TABLE'
    });
    this.logsTable = lambdaToDynamoDb.dynamoTable;

    const cfnDynamoTable = <CfnTable>this.logsTable.node.defaultChild;
    cfnDynamoTable.cfnOptions.deletionPolicy = CfnDeletionPolicy.DELETE;

    const infoLogsProps: IotToSqsProps = {
      existingQueueObj: sqsToLambda.sqsQueue,
      iotTopicRuleProps: {
        topicRulePayload: {
          actions: [],
          awsIotSqlVersion: '2016-03-23',
          description: 'Processing connection info logs',
          sql: this.sql.replace('{type}', 'info')
        }
      }
    };
    new IotToSqs(this, 'InfoLogsRule', infoLogsProps);

    const errorLogsProps: IotToSqsProps = {
      existingQueueObj: sqsToLambda.sqsQueue,
      iotTopicRuleProps: {
        topicRulePayload: {
          actions: [],
          awsIotSqlVersion: '2016-03-23',
          description: 'Processing connection error logs',
          sql: this.sql.replace('{type}', 'error')
        }
      }
    };
    new IotToSqs(this, 'ErrorLogsRule', errorLogsProps);

    // cdk-nag suppressions
    NagSuppressions.addResourceSuppressions(
      sqsToLambda,
      [
        { id: 'AwsSolutions-IAM5', reason: 'It does not allow wildcard permissions' },
        { id: 'AwsSolutions-SQS3', reason: 'deadLetterQueue is the dead letter queue.' }
      ],
      true
    );
  }
}
