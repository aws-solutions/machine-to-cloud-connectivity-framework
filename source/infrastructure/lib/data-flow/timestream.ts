// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { KinesisStreamsToLambda } from '@aws-solutions-constructs/aws-kinesisstreams-lambda';
import { CfnCondition, CfnDeletionPolicy, Duration, Fn, CustomResource, CfnCustomResource, Aws } from 'aws-cdk-lib';
import { Effect, Policy, PolicyStatement, Role } from 'aws-cdk-lib/aws-iam';
import { Code, Runtime } from 'aws-cdk-lib/aws-lambda';
import { CfnDatabase, CfnTable } from 'aws-cdk-lib/aws-timestream';
import { IBucket } from 'aws-cdk-lib/aws-s3';
import { NagSuppressions } from 'cdk-nag';
import { Construct } from 'constructs';
import { addCfnSuppressRules } from '../../utils/utils';

export interface TimestreamConstructProps {
  readonly existingDatabaseName: string;
  readonly solutionConfig: {
    loggingLevel: string;
    solutionId: string;
    solutionVersion: string;
    sourceCodeBucket: IBucket;
    sourceCodePrefix: string;
    uuid: string;
  };
  readonly customResourcesFunctionArn: string;
  readonly shouldTeardownData: CfnCondition;
}

/**
 * Creates a Lambda function, Kinesis Data Stream streams, Timestream database and table, and other related resources.
 */
export class TimestreamConstruct extends Construct {
  public kinesisStreamArn: string;
  public kinesisStreamName: string;
  public timestreamDatabaseTable: string;

  constructor(scope: Construct, id: string, props: TimestreamConstructProps) {
    super(scope, id);

    const sourceCodeBucket = props.solutionConfig.sourceCodeBucket;
    const sourceCodePrefix = props.solutionConfig.sourceCodePrefix;

    const createTimestreamDatabaseCondition = new CfnCondition(this, 'CreateTimestreamDatabase', {
      expression: Fn.conditionEquals(props.existingDatabaseName, '')
    });

    const timestreamDatabase = new CfnDatabase(this, 'Database', {
      databaseName: `${Aws.STACK_NAME}-${props.solutionConfig.uuid}`
    });
    timestreamDatabase.cfnOptions.condition = createTimestreamDatabaseCondition;
    timestreamDatabase.cfnOptions.deletionPolicy = CfnDeletionPolicy.RETAIN;

    const timestreamDatabaseName = Fn.conditionIf(
      createTimestreamDatabaseCondition.logicalId,
      timestreamDatabase.ref,
      props.existingDatabaseName
    ).toString();

    const teardownTimestreamDatabase = new CustomResource(this, 'teardownTimestreamDatabase', {
      serviceToken: props.customResourcesFunctionArn,
      properties: {
        Resource: 'DeleteTimestreamDatabase',
        DatabaseName: timestreamDatabaseName
      }
    });
    const cfnTeardownGreengrassResourcesBucket = <CfnCustomResource>teardownTimestreamDatabase.node.defaultChild;
    cfnTeardownGreengrassResourcesBucket.cfnOptions.condition = props.shouldTeardownData;

    const timestreamTable = new CfnTable(this, 'Table', {
      databaseName: timestreamDatabaseName,
      retentionProperties: {
        MemoryStoreRetentionPeriodInHours: 24 * 90,
        MagneticStoreRetentionPeriodInDays: 365
      }
    });
    const timestreamTableName = timestreamTable.getAtt('Name').toString();
    const timestreamTableArn = timestreamTable.getAtt('Arn').toString();
    timestreamTable.cfnOptions.deletionPolicy = CfnDeletionPolicy.RETAIN;

    const kinesisStreamsToLambda = new KinesisStreamsToLambda(this, 'KinesisLambda', {
      lambdaFunctionProps: {
        code: Code.fromBucket(sourceCodeBucket, `${sourceCodePrefix}/timestream-writer.zip`),
        description: 'Machine to Cloud Connectivity Framework Timestream data writer function',
        environment: {
          LOGGING_LEVEL: props.solutionConfig.loggingLevel,
          SOLUTION_ID: props.solutionConfig.solutionId,
          SOLUTION_VERSION: props.solutionConfig.solutionVersion,
          TIMESTREAM_DATABASE: timestreamDatabaseName,
          TIMESTREAM_TABLE: timestreamTableName
        },
        handler: 'timestream-writer/index.handler',
        runtime: Runtime.NODEJS_16_X,
        timeout: Duration.seconds(30)
      }
    });

    const lambdaFunctionPolicy = new Policy(this, 'TimestreamPolicy', {
      statements: [
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: ['timestream:WriteRecords'],
          resources: [timestreamTableArn]
        }),
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: ['timestream:DescribeEndpoints'],
          resources: ['*']
        })
      ]
    });

    addCfnSuppressRules(lambdaFunctionPolicy, [
      { id: 'W12', reason: 'timestream:DescribeEndpoints cannot have specific resources.' }
    ]);

    const lambdaFunctionRole = <Role>kinesisStreamsToLambda.lambdaFunction.role;
    lambdaFunctionRole.attachInlinePolicy(lambdaFunctionPolicy);

    this.kinesisStreamArn = kinesisStreamsToLambda.kinesisStream.streamArn;
    this.kinesisStreamName = kinesisStreamsToLambda.kinesisStream.streamName;
    this.timestreamDatabaseTable = `${timestreamDatabaseName}.${timestreamTableName}`;

    // cdk-nag suppressions
    NagSuppressions.addResourceSuppressions(
      kinesisStreamsToLambda,
      [
        { id: 'AwsSolutions-IAM5', reason: 'It does not allow wildcard permissions.' },
        { id: 'AwsSolutions-SQS3', reason: 'SqsDlqQueue is the dead letter queue.' }
      ],
      true
    );
    NagSuppressions.addResourceSuppressions(timestreamDatabase, [
      { id: 'AwsSolutions-TS3', reason: 'The default KMS is used by default.' }
    ]);
    NagSuppressions.addResourceSuppressions(lambdaFunctionPolicy, [
      { id: 'AwsSolutions-IAM5', reason: 'timestream:DescribeEndpoints cannot have specific resources.' }
    ]);
  }
}
