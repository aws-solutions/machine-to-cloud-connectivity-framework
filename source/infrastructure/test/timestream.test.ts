// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { Template } from 'aws-cdk-lib/assertions';
import { CfnCondition, Stack, assertions, aws_s3 as s3, aws_timestream as timestream } from 'aws-cdk-lib';
import { TimestreamConstruct } from '../lib/data-flow/timestream';

const loggingLevel = 'ERROR';
const solutionId = 'SO0070-Test';
const solutionVersion = 'v0.0.1-test';
const sourceCodePrefix = `machine-to-cloud-connectivity-framework/${solutionVersion}`;
const uuid = 'test-uuid';

test('M2C2 Timestream test snapshot and default values', () => {
  const stack = new Stack();
  const sourceCodeBucket = s3.Bucket.fromBucketName(stack, 'SourceCodeBucket', 'test-bucket-region');
  const timestream = new TimestreamConstruct(stack, 'TestTimestream', {
    existingDatabaseName: '',
    solutionConfig: {
      loggingLevel,
      solutionId,
      solutionVersion,
      sourceCodeBucket,
      sourceCodePrefix,
      uuid
    },
    customResourcesFunctionArn: '',
    shouldTeardownData: new CfnCondition(stack, 'TestCondition')
  });
  const table = <timestream.CfnTable>timestream.node.findChild('Table');
  table.overrideLogicalId('TestTable');

  expect(timestream.kinesisStreamName).toBeDefined();
  expect(timestream.timestreamDatabaseTable).toBeDefined();
  Template.fromStack(stack).hasResourceProperties('AWS::Lambda::Function', {
    Environment: {
      Variables: {
        LOGGING_LEVEL: 'ERROR',
        SOLUTION_ID: solutionId,
        SOLUTION_VERSION: solutionVersion
      }
    },
    Handler: 'timestream-writer/index.handler',
    Runtime: 'nodejs18.x',
    Timeout: 30
  });
  Template.fromStack(stack).hasResourceProperties('AWS::IAM::Policy', {
    PolicyDocument: {
      Statement: [
        {
          Action: 'timestream:WriteRecords',
          Effect: 'Allow',
          Resource: {
            'Fn::GetAtt': ['TestTable', 'Arn']
          }
        },
        {
          Action: 'timestream:DescribeEndpoints',
          Effect: 'Allow',
          Resource: '*'
        }
      ]
    }
  });

  const template = assertions.Template.fromStack(stack);
  template.resourceCountIs('AWS::Kinesis::Stream', 1);
});

test('M2C2 Timestream test when Timestream database name parameter is provided', () => {
  const stack = new Stack();
  const sourceCodeBucket = s3.Bucket.fromBucketName(stack, 'SourceCodeBucket', 'test-bucket-region');
  const timestream = new TimestreamConstruct(stack, 'TestTimestream', {
    existingDatabaseName: 'test-database',
    solutionConfig: {
      loggingLevel,
      solutionId,
      solutionVersion,
      sourceCodeBucket,
      sourceCodePrefix,
      uuid
    },
    customResourcesFunctionArn: '',
    shouldTeardownData: new CfnCondition(stack, 'TestCondition')
  });
  const database = <timestream.CfnDatabase>timestream.node.findChild('Database');
  database.overrideLogicalId('TestDatabase');
  const condition = <CfnCondition>timestream.node.findChild('CreateTimestreamDatabase');
  condition.overrideLogicalId('TestCondition');

  Template.fromStack(stack).hasResourceProperties('AWS::Timestream::Table', {
    DatabaseName: {
      'Fn::If': [
        'TestCondition',
        {
          Ref: 'TestDatabase'
        },
        'test-database'
      ]
    }
  });
});

test('M2C2 Timestream test when Timestream database name parameter is empty', () => {
  const stack = new Stack();
  const sourceCodeBucket = s3.Bucket.fromBucketName(stack, 'SourceCodeBucket', 'test-bucket-region');
  const timestream = new TimestreamConstruct(stack, 'TestTimestream', {
    existingDatabaseName: '',
    solutionConfig: {
      loggingLevel,
      solutionId,
      solutionVersion,
      sourceCodeBucket,
      sourceCodePrefix,
      uuid
    },
    customResourcesFunctionArn: '',
    shouldTeardownData: new CfnCondition(stack, 'TestCondition')
  });
  const database = <timestream.CfnDatabase>timestream.node.findChild('Database');
  database.overrideLogicalId('TestDatabase');
  const condition = <CfnCondition>timestream.node.findChild('CreateTimestreamDatabase');
  condition.overrideLogicalId('TestCondition');

  Template.fromStack(stack).hasResourceProperties('AWS::Timestream::Table', {
    DatabaseName: {
      'Fn::If': [
        'TestCondition',
        {
          Ref: 'TestDatabase'
        },
        ''
      ]
    }
  });
});
