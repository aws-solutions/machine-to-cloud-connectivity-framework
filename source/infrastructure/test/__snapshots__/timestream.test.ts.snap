// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`M2C2 Timestream test snapshot and default values 1`] = `
Object {
  "Conditions": Object {
    "TestTimestreamCreateTimestreamDatabaseDCDEE72E": Object {
      "Fn::Equals": Array [
        "",
        "",
      ],
    },
  },
  "Resources": Object {
    "TestTable": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "DatabaseName": Object {
          "Fn::If": Array [
            "TestTimestreamCreateTimestreamDatabaseDCDEE72E",
            Object {
              "Ref": "TestTimestreamDatabaseD9B9FB32",
            },
            "",
          ],
        },
        "RetentionProperties": Object {
          "MagneticStoreRetentionPeriodInDays": 365,
          "MemoryStoreRetentionPeriodInHours": 2160,
        },
      },
      "Type": "AWS::Timestream::Table",
    },
    "TestTimestreamDatabaseD9B9FB32": Object {
      "Condition": "TestTimestreamCreateTimestreamDatabaseDCDEE72E",
      "DeletionPolicy": "Retain",
      "Metadata": Object {
        "cdk_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "AwsSolutions-TS3",
              "reason": "The default KMS is used by default.",
            },
          ],
        },
      },
      "Type": "AWS::Timestream::Database",
    },
    "TestTimestreamKinesisLambdaKinesisStreamD9ACCC8B": Object {
      "Metadata": Object {
        "cdk_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "AwsSolutions-IAM5",
              "reason": "It does not allow wildcard permissions.",
            },
            Object {
              "id": "AwsSolutions-SQS3",
              "reason": "SqsDlqQueue is the dead letter queue.",
            },
          ],
        },
      },
      "Properties": Object {
        "RetentionPeriodHours": 24,
        "ShardCount": 1,
        "StreamEncryption": Object {
          "EncryptionType": "KMS",
          "KeyId": "alias/aws/kinesis",
        },
        "StreamModeDetails": Object {
          "StreamMode": "PROVISIONED",
        },
      },
      "Type": "AWS::Kinesis::Stream",
    },
    "TestTimestreamKinesisLambdaKinesisStreamGetRecordsIteratorAgeAlarm3BF96A2B": Object {
      "Metadata": Object {
        "cdk_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "AwsSolutions-IAM5",
              "reason": "It does not allow wildcard permissions.",
            },
            Object {
              "id": "AwsSolutions-SQS3",
              "reason": "SqsDlqQueue is the dead letter queue.",
            },
          ],
        },
      },
      "Properties": Object {
        "AlarmDescription": "Consumer Record Processing Falling Behind, there is risk for data loss due to record expiration.",
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "EvaluationPeriods": 1,
        "MetricName": "GetRecords.IteratorAgeMilliseconds",
        "Namespace": "AWS/Kinesis",
        "Period": 300,
        "Statistic": "Maximum",
        "Threshold": 2592000,
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "TestTimestreamKinesisLambdaKinesisStreamReadProvisionedThroughputExceededAlarm84FAAD02": Object {
      "Metadata": Object {
        "cdk_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "AwsSolutions-IAM5",
              "reason": "It does not allow wildcard permissions.",
            },
            Object {
              "id": "AwsSolutions-SQS3",
              "reason": "SqsDlqQueue is the dead letter queue.",
            },
          ],
        },
      },
      "Properties": Object {
        "AlarmDescription": "Consumer Application is Reading at a Slower Rate Than Expected.",
        "ComparisonOperator": "GreaterThanThreshold",
        "EvaluationPeriods": 1,
        "MetricName": "ReadProvisionedThroughputExceeded",
        "Namespace": "AWS/Kinesis",
        "Period": 300,
        "Statistic": "Average",
        "Threshold": 0,
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "TestTimestreamKinesisLambdaLambdaFunctionB7668DD3": Object {
      "DependsOn": Array [
        "TestTimestreamKinesisLambdaLambdaFunctionServiceRoleDefaultPolicy469FCFC6",
        "TestTimestreamKinesisLambdaLambdaFunctionServiceRole1DFBA77A",
      ],
      "Metadata": Object {
        "cdk_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "AwsSolutions-IAM5",
              "reason": "It does not allow wildcard permissions.",
            },
            Object {
              "id": "AwsSolutions-SQS3",
              "reason": "SqsDlqQueue is the dead letter queue.",
            },
          ],
        },
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W58",
              "reason": "Lambda functions has the required permission to write CloudWatch Logs. It uses custom policy instead of arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole with tighter permissions.",
            },
            Object {
              "id": "W89",
              "reason": "This is not a rule for the general case, just for specific use cases/industries",
            },
            Object {
              "id": "W92",
              "reason": "Impossible for us to define the correct concurrency for clients",
            },
          ],
        },
      },
      "Properties": Object {
        "Code": Object {
          "S3Bucket": "test-bucket-region",
          "S3Key": "machine-to-cloud-connectivity-framework/v0.0.1-test/timestream-writer.zip",
        },
        "Description": "Machine to Cloud Connectivity Framework Timestream data writer function",
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "LOGGING_LEVEL": "ERROR",
            "SOLUTION_ID": "SO0070-Test",
            "SOLUTION_VERSION": "v0.0.1-test",
            "TIMESTREAM_DATABASE": Object {
              "Fn::If": Array [
                "TestTimestreamCreateTimestreamDatabaseDCDEE72E",
                Object {
                  "Ref": "TestTimestreamDatabaseD9B9FB32",
                },
                "",
              ],
            },
            "TIMESTREAM_TABLE": Object {
              "Fn::GetAtt": Array [
                "TestTable",
                "Name",
              ],
            },
          },
        },
        "Handler": "timestream-writer/index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "TestTimestreamKinesisLambdaLambdaFunctionServiceRole1DFBA77A",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 30,
        "TracingConfig": Object {
          "Mode": "Active",
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "TestTimestreamKinesisLambdaLambdaFunctionKinesisEventSourceTestTimestreamKinesisLambdaKinesisStreamE555AD84EA3892B3": Object {
      "Metadata": Object {
        "cdk_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "AwsSolutions-IAM5",
              "reason": "It does not allow wildcard permissions.",
            },
            Object {
              "id": "AwsSolutions-SQS3",
              "reason": "SqsDlqQueue is the dead letter queue.",
            },
          ],
        },
      },
      "Properties": Object {
        "BatchSize": 100,
        "BisectBatchOnFunctionError": true,
        "DestinationConfig": Object {
          "OnFailure": Object {
            "Destination": Object {
              "Fn::GetAtt": Array [
                "TestTimestreamKinesisLambdaSqsDlqQueue965BC5DD",
                "Arn",
              ],
            },
          },
        },
        "EventSourceArn": Object {
          "Fn::GetAtt": Array [
            "TestTimestreamKinesisLambdaKinesisStreamD9ACCC8B",
            "Arn",
          ],
        },
        "FunctionName": Object {
          "Ref": "TestTimestreamKinesisLambdaLambdaFunctionB7668DD3",
        },
        "MaximumRecordAgeInSeconds": 86400,
        "MaximumRetryAttempts": 500,
        "StartingPosition": "TRIM_HORIZON",
      },
      "Type": "AWS::Lambda::EventSourceMapping",
    },
    "TestTimestreamKinesisLambdaLambdaFunctionServiceRole1DFBA77A": Object {
      "Metadata": Object {
        "cdk_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "AwsSolutions-IAM5",
              "reason": "It does not allow wildcard permissions.",
            },
            Object {
              "id": "AwsSolutions-SQS3",
              "reason": "SqsDlqQueue is the dead letter queue.",
            },
          ],
        },
      },
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "logs:CreateLogGroup",
                    "logs:CreateLogStream",
                    "logs:PutLogEvents",
                  ],
                  "Effect": "Allow",
                  "Resource": Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        "arn:",
                        Object {
                          "Ref": "AWS::Partition",
                        },
                        ":logs:",
                        Object {
                          "Ref": "AWS::Region",
                        },
                        ":",
                        Object {
                          "Ref": "AWS::AccountId",
                        },
                        ":log-group:/aws/lambda/*",
                      ],
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "LambdaFunctionServiceRolePolicy",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TestTimestreamKinesisLambdaLambdaFunctionServiceRoleDefaultPolicy469FCFC6": Object {
      "Metadata": Object {
        "cdk_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "AwsSolutions-IAM5",
              "reason": "It does not allow wildcard permissions.",
            },
            Object {
              "id": "AwsSolutions-SQS3",
              "reason": "SqsDlqQueue is the dead letter queue.",
            },
          ],
        },
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W12",
              "reason": "Lambda needs the following minimum required permissions to send trace data to X-Ray and access ENIs in a VPC.",
            },
          ],
        },
      },
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "xray:PutTraceSegments",
                "xray:PutTelemetryRecords",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "kinesis:DescribeStreamSummary",
                "kinesis:GetRecords",
                "kinesis:GetShardIterator",
                "kinesis:ListShards",
                "kinesis:SubscribeToShard",
                "kinesis:DescribeStream",
                "kinesis:ListStreams",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "TestTimestreamKinesisLambdaKinesisStreamD9ACCC8B",
                  "Arn",
                ],
              },
            },
            Object {
              "Action": Array [
                "sqs:SendMessage",
                "sqs:GetQueueAttributes",
                "sqs:GetQueueUrl",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "TestTimestreamKinesisLambdaSqsDlqQueue965BC5DD",
                  "Arn",
                ],
              },
            },
            Object {
              "Action": "kinesis:DescribeStream",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "TestTimestreamKinesisLambdaKinesisStreamD9ACCC8B",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TestTimestreamKinesisLambdaLambdaFunctionServiceRoleDefaultPolicy469FCFC6",
        "Roles": Array [
          Object {
            "Ref": "TestTimestreamKinesisLambdaLambdaFunctionServiceRole1DFBA77A",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TestTimestreamKinesisLambdaSqsDlqQueue965BC5DD": Object {
      "DeletionPolicy": "Delete",
      "Metadata": Object {
        "cdk_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "AwsSolutions-IAM5",
              "reason": "It does not allow wildcard permissions.",
            },
            Object {
              "id": "AwsSolutions-SQS3",
              "reason": "SqsDlqQueue is the dead letter queue.",
            },
          ],
        },
      },
      "Properties": Object {
        "KmsMasterKeyId": "alias/aws/sqs",
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "TestTimestreamKinesisLambdaSqsDlqQueuePolicy5A20611D": Object {
      "Metadata": Object {
        "cdk_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "AwsSolutions-IAM5",
              "reason": "It does not allow wildcard permissions.",
            },
            Object {
              "id": "AwsSolutions-SQS3",
              "reason": "SqsDlqQueue is the dead letter queue.",
            },
          ],
        },
      },
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "sqs:DeleteMessage",
                "sqs:ReceiveMessage",
                "sqs:SendMessage",
                "sqs:GetQueueAttributes",
                "sqs:RemovePermission",
                "sqs:AddPermission",
                "sqs:SetQueueAttributes",
              ],
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "TestTimestreamKinesisLambdaSqsDlqQueue965BC5DD",
                  "Arn",
                ],
              },
              "Sid": "QueueOwnerOnlyAccess",
            },
            Object {
              "Action": "SQS:*",
              "Condition": Object {
                "Bool": Object {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": Object {
                "AWS": "*",
              },
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "TestTimestreamKinesisLambdaSqsDlqQueue965BC5DD",
                  "Arn",
                ],
              },
              "Sid": "HttpsOnly",
            },
          ],
          "Version": "2012-10-17",
        },
        "Queues": Array [
          Object {
            "Ref": "TestTimestreamKinesisLambdaSqsDlqQueue965BC5DD",
          },
        ],
      },
      "Type": "AWS::SQS::QueuePolicy",
    },
    "TestTimestreamTimestreamPolicy68B07AB6": Object {
      "Metadata": Object {
        "cdk_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "AwsSolutions-IAM5",
              "reason": "timestream:DescribeEndpoints cannot have specific resources.",
            },
          ],
        },
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W12",
              "reason": "timestream:DescribeEndpoints cannot have specific resources.",
            },
          ],
        },
      },
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "timestream:WriteRecords",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "TestTable",
                  "Arn",
                ],
              },
            },
            Object {
              "Action": "timestream:DescribeEndpoints",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TestTimestreamTimestreamPolicy68B07AB6",
        "Roles": Array [
          Object {
            "Ref": "TestTimestreamKinesisLambdaLambdaFunctionServiceRole1DFBA77A",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
